### ISSUE TITLE
Drop or deprecate support for environment variable MYSQL_PWD

### ISSUE THREAD

**Issue by tsroten** [dbcli/mycli#435]
The MySQL documentation [strongly recommends against](https://dev.mysql.com/doc/refman/5.5/en/password-security-user.html) using this feature:

> This method of specifying your MySQL password must be considered extremely insecure and should not be used. Some versions of ps include an option to display the environment of running processes. On some systems, if you set MYSQL_PWD, your password is exposed to any other user who runs ps. Even on systems without such a version of ps, it is unwise to assume that there are no other methods by which users can examine process environments.

Or, we could deprecate it by issuing a warning to the user if that environment variable is set.

**Comment by tsroten** [comment]
@amjith I'm curious for your thoughts on this.

**Comment by amjith** [comment]
I thought using env vars was one of the recommended ways to pass in credentials to a microservices. I know that docker uses it all the time and a lot of application servers use the env vars to get the credentials to connect to database. 

I'm not so sure this is such a security risk. But I'm also not an expert in that area. 

I'm fine printing a warning as part of the startup, we can even make that RED. Something that reads, "You're using MYSQL_PWD, this is considered insecure. Please avoid using this feature. More details: <url>".

**Comment by meeuw** [comment]
I think the mysql documentation is a bit exaggerating, I know more systems that are sharing command arguments in `ps` output than the environment variables. 

Docker is limited to Linux only, and as far as I know the environment variables are safe on Linux.

If we start warning about `MYSQL_PWD` we also should warn about `-p` and warn if `.my.cnf` is readable for other users (like openssh's `.ssh/id_rsa`). It would be nice to have a configuration option to disable the warnings (for a single user system).

**Comment by tsroten** [comment]
@meeuw @amjith I agree that it seems a bit overboard. What if we simply deprecate it through the documentation. In other words, we make it clear what the recommended way to specify passwords is.

**Comment by meeuw** [comment]
Just to be sure; the recommended way is by using `mysql_config_editor` with a master password?

As I think of it; it would be nice if we could write an encrypted password using mycli...

**Comment by tsroten** [comment]
@meeuw  The problem with `mysql_config_editor` is that you do not control the "master password". If you use `mysql_config_editor` and I get a hold of your login path file, I can decrypt it using `mysql_config_editor` or mycli.

It's been pointed out a few times that mycli's support of `mysql_config_editor` files is not a security feature. It's a convenience feature for those people who use mysql.

Currently, the best option for secure password handling in mycli is to type it in manually each time.

I've been thinking that we should consider adopting back ticks to execute shell code in the `myclirc`. Each OS has different built-in ways to handle encrypted passwords. On macOS, it's `security`:

```
$ security find-generic-password -gw -a root -s localhost
```

That's how mutt and offlineimap get my password for my email.

So, in `myclirc`, it would be:
```
[client]
password=`security find-generic-password -gw -a root -s localhost`
```

Other OS's have things like GNOME Keyring or [Secret Service](http://standards.freedesktop.org/secret-service/).

Whatever we do, we definitely don't want to roll our own. The issue of encrypted passwords has already been solved by each operating system. We need to tap into that, like other CLI apps do. Another possibility for this is [keyring](https://github.com/jaraco/keyring).

**Comment by meeuw** [comment]
@tsroten you are completely right, maybe we could go a bit futher like Google Chrome uses the operating system default keyring, maybe implement this in cli_helpers?

**Comment by tsroten** [comment]
@meeuw That's a good idea. We could have a CLI Helper module that wraps around [`keyring`](https://github.com/jaraco/keyring). I like the idea of using an existing library behind the scenes, since it would be laborious to support each OS's keyring on our own.

This would be a good thing to add into mycli once I move all the config reading/writing code out of mycli -- we could look up the password in the keyring if the user does not supply a password.

**Comment by duzun** [comment]
What are the options for handling the password in automated environments/scenarios (say docker)?

I'm aware of three options:
1. CLI argument like `mysql -u"root" -p"secret" -h"mydbhost.com"`
2. env var `export MYSQL_PWD=secret; mysql -u"root" -h"mydbhost.com"`
3. config file `~/.my.cnf`, with `chmod 600 ~/.my.cnf`, then just `mysql -h"mydbhost.com"`
    3.5. a custom extra config file `/any/path/my-custom-credentials.cnf` also `chmod`ed to `600`, used like `mysql --defaults-extra-file=/any/path/my-custom-credentials.cnf -h"mydbhost.com"`

```ini
; Sample .my.cnf or my-custom-credentials.cnf file
[mysqldump]
user=root
password=secret
```

Option `1.` is the worst one for obvious reasons.
Option `2.` is my vaforite so far, cause only root and the running user of the command can access environment variables of `mysql` process, so basically it is not worse than option `3.` in terms of security, but has the advantage of being more ephemeral (no process -> no password info) and easier to pass around (no need co create a config file).

Am I missing anything?

**Comment by chutzimir** [comment]
> I think the mysql documentation is a bit exaggerating, I know more systems that are sharing command arguments in `ps` output than the environment variables.

I think the versions of `ps` which can show the environment variables (e.g. `ps auxe`) only do that if the `ps` process itself can access the environment information (the user itself, or root), while command line arguments are visible to anyone. Cannot speak for all OS flavors out there, but it is true for Linux.

This thread was referenced from [a similar question on StackOverflow](https://stackoverflow.com/a/34670902/1213346) so it would be nice to get some kind of conclusive response here which can be shared with the community.

**Comment by tsroten** [comment]
@chutzimir @duzun Thanks for your input! I believe our current plan is to leave the support for the environment variable as is due to its convenience and wide-spread use. However, we plan to add support for encrypted passwords via [`keyring`](https://github.com/jaraco/keyring) like pgcli has done. This will supersede our support for MySQL's [encrypted login path file](https://dev.mysql.com/doc/refman/5.6/en/mysql-config-editor.html), which is trivial to decrypt.

But none of this is really applicable to automated environments as mycli is really not the best tool for the job, the official MySQL client is. Mycli is great for human interaction.

**Comment by chutzimir** [comment]
> But none of this is really applicable to automated environments as mycli is really not the best tool for the job, the official MySQL client is. Mycli is great for human interaction.

OMG, just now I noticed that this discussion is about `mycli` which I had never heard about. Always assumed we are talking about the `mysql` tool. Thanks for the update.

**Comment by rolandwalker** [comment]
Closing as this is resolved in favor of keeping support for the environment variable.
